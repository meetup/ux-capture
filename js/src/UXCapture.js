import ExpectedMark from "./ExpectedMark";
import View from "./View";

const NOOP = () => {};

export default class UXCapture {
  constructor() {
    this.onMark = NOOP;
    this.onMeasure = NOOP;

    window.UX = this;
  }

  /**
   * Sets expected marks and corresponding zones for current view
   *
   * @TODO re-evaluate if we should allow multiple executions of this method
   */
  expect(zoneConfigs) {
    if (typeof zoneConfigs === "undefined" || !Array.isArray(zoneConfigs)) {
      throw new Error("Must provide an array of zone configuration objects");
    }

    // create a view object for initial, server-side rendered page view
    const pageView = new View({
      // calling currently configured onMark & onMeasure handlers inside View's callbacks
      // allows making UX.config() call after UX.expect() call
      onMark: mark => {
        // console.log(this.onMark);
        this.onMark(mark);
      },
      onMeasure: measure => {
        // console.log(this.onMeasure);
        this.onMeasure(measure);
      },
      zoneConfigs
    });
  }

  mark(name) {
    const mark = ExpectedMark.get(name);

    if (mark) {
      mark.record();
    }
  }

  config(configuration) {
    const { onMark = NOOP, onMeasure = NOOP } = configuration;
    Object.assign(this, { onMark, onMeasure });
  }
}
