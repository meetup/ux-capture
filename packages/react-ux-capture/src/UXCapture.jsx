// @flow
import * as React from 'react';
import UXCaptureCreate from './UXCaptureCreate';
import UXCaptureStartView from './UXCaptureStartView';

type Props = {|
	destinationVerified?: Array<string>,
	primaryContentDisplayed?: Array<string>,
	primaryActionAvailable?: Array<string>,
	secondaryContentDisplayed?: Array<string>,
|};

export const onMark = (mark: string) => {
	if (
		!(
			window.newrelic &&
			window.performance &&
			window.performance.getEntriesByType
		)
	) {
		return;
	}

	const performanceMark = window.performance
		.getEntriesByType('mark')
		.find(entry => entry.name === mark);

	// Set a marker in the trace details
	window.newrelic.addToTrace({
		name: mark,
		start: window.performance.timing.navigationStart + performanceMark.startTime, // this is an epoch ms timestamp
		type: 'UX Capture mark',
	});
};

export const onMeasure = (measure: string) => {
	if (
		!(
			window.newrelic &&
			window.performance &&
			window.performance.getEntriesByType
		)
	) {
		return;
	}

	const performanceMeasure = window.performance
		.getEntriesByType('measure')
		.find(entry => entry.name === measure);

	//  Add a custom attribute to the PageView & BrowserInteraction events in Insights
	window.newrelic.setCustomAttribute(measure, performanceMeasure.duration);
};

export default ({
	destinationVerified,
	primaryContentDisplayed,
	primaryActionAvailable,
	secondaryContentDisplayed,
}: Props) => (
	<React.Fragment>
		<UXCaptureCreate onMark={onMark} onMeasure={onMeasure} />
		<UXCaptureStartView
			destinationVerified={destinationVerified}
			primaryContentDisplayed={primaryContentDisplayed}
			primaryActionAvailable={primaryActionAvailable}
			secondaryContentDisplayed={secondaryContentDisplayed}
		/>
	</React.Fragment>
);
