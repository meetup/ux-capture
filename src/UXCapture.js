import ExpectedMark from './ExpectedMark';
import PageView from './PageView';
import InteractiveView from './InteractiveView';

const NOOP = () => {};

let _onMark;
let _onMeasure;
let _view;

const UXCapture = {
	_clearMarksAndMeasures: () => {
		window.performance.clearMarks();
		window.performance.clearMeasures();
	},

	/**
	 * Sets `onMark` and `onMeasure` callbacks on UXCapture singleton
	 * Also resets the view (supposed to be called once per page anyway)
	 *
	 * @param {object} config
	 */
	create: config => {
		_onMark = config.onMark || NOOP;
		_onMeasure = config.onMeasure || NOOP;
		_view = null;
	},

	/**
	 * Creates a new View instance.
	 *
	 * @param {object} zoneConfigs
	 */
	startView: zoneConfigs => {
		const viewProps = {
			onMark: _onMark,
			onMeasure: _onMeasure,
			zoneConfigs,
		};

		// if we never had views before, it means we are in page view mode
		if (_view) {
			_view = new InteractiveView(viewProps);
		} else {
			_view = new PageView(viewProps);
		}
	},

	/**
	 * Updates current view instance with new zones & marks
	 *
	 * @param {object} zoneConfigs
	 */
	updateView: zoneConfigs => {
		if (!_view) {
			window.console.error(
				'[Error] No view to update. Call UXCapture.startView() before UXCapture.updateView()'
			);
			return;
		}

		_view.update(zoneConfigs);
	},

	startTransition: () => {
		UXCapture._clearMarksAndMeasures();

		_view.startTransition();
	},

	/**
	 * Creates marks on UserTiming timeline.
	 *
	 * `waitForNextPaint` should be set to false if your code is not
	 * expected to change any visual elements and trigger repaints.
	 *
	 * Example: click event handlers on elements are necessary
	 * for them to be interactive but don't affect the element's appearance.
	 *
	 * @param {string} name
	 * @param {boolean} waitForNextPaint
	 */
	mark: (name, waitForNextPaint = true) => {
		const mark = ExpectedMark.get(name);

		if (mark) {
			if (waitForNextPaint) {
				mark.waitForNextPaintAndRecord();
			} else {
				mark.record();
			}
		}
	},
};

export default UXCapture;
