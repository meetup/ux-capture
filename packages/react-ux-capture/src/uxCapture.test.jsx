import React from 'react';
import { shallow } from 'enzyme';
import UXCapture from './UXCapture';

const MOCK_PROPS = {
	destinationVerfied: ['ux-destination-verified'],
	primaryContentDisplayed: ['ux-primary-content-1', 'ux-primary-content-2'],
	primaryActionAvailable: ['ux-primary-action-available'],
	secondaryContentDisplayed: [
		'ux-secondary-content-1',
		'ux-secondary-content-2',
		'ux-secondary-content-3',
	],
};

const renderComponent = (props = MOCK_PROPS) =>
	shallow(<UXCapture {...props} />);

describe('UXCapture', () => {
	it('renders correct markup when all props are defined', () => {
		expect(renderComponent()).toMatchSnapshot();
	});
});
