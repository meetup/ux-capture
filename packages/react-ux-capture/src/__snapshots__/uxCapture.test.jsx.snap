// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UXCapture renders correct markup when all props are defined 1`] = `
<Fragment>
  <_default
    onMark={[Function]}
    onMeasure={[Function]}
  />
  <UXCaptureStartView
    primaryActionAvailable={
      Array [
        "ux-primary-action-available",
      ]
    }
    primaryContentDisplayed={
      Array [
        "ux-primary-content-1",
        "ux-primary-content-2",
      ]
    }
    secondaryContentDisplayed={
      Array [
        "ux-secondary-content-1",
        "ux-secondary-content-2",
        "ux-secondary-content-3",
      ]
    }
  />
</Fragment>
`;
